Infrastructure-Scripts/Set-AppRoleAssignments/tools/Helpers.psm1

```diff
@@ -46,6 +45,35 @@ function New-AppRegistration {
     az ad sp create --id $IdentifierUri --output none 2>$null
 }
 
+function Set-AzureCLIAccess {
+    Param(
+        [Parameter(Mandatory = $true)]
+        [String]$IdentifierUri,
+        [Parameter(Mandatory = $true)]
+        [String]$ServicePrincipalObjectId
+    )
+
+    #Authorize Azure CLI to call app registration and acquire a token
+    #https://docs.microsoft.com/en-us/graph/api/resources/preauthorizedapplication?view=graph-rest-1.0
+    $AppRegistrationObject = (az ad app show --id $IdentifierUri | ConvertFrom-Json)
+    $MicrosoftGraphRequestParameters =
+    "--method", "patch",
+    "--uri", "https://graph.microsoft.com/v1.0/applications/$($AppRegistrationObject.ObjectId)",
+    "--headers", "Content-Type=application/json",
+    "--body", "{api:{preAuthorizedApplications:[{appId:'04b07795-8ddb-461a-bbee-02f9e1bf7b46',delegatedPermissionIds:['$($AppRegistrationObject.Oauth2Permissions.id)']}]}}"
+
+    az rest @MicrosoftGraphRequestParameters
+
+    #Apply User Assignment required so only authorized users can acquire a token
+    #https://docs.microsoft.com/en-us/graph/api/serviceprincipal-update?view=graph-rest-1.0&tabs=http
+    $MicrosoftGraphRequestParameters =
+    "--method", "patch",
+    "--uri", "https://graph.microsoft.com/v1.0/servicePrincipals/$ServicePrincipalObjectId",
+    "--headers", "Content-Type=application/json",
+    "--body", "{appRoleAssignmentRequired : true}"
+    az rest @MicrosoftGraphRequestParameters
```

Comment:
Can this be the first block, before the Azure CLI is authorized so there's not chance of the script failing part way through and User Assignment not required?